### DISCUSSION TITLE
.ConfigureAwait(false) missing in DataPortal access async methods

### DISCUSSION THREAD

**Discussion by bil-bal** [MarimerLLC/csla#3879] [Questions]
Hi, we had a case where the UI (using WPF) was blocking for a few seconds at the end of fetching a `BusinessListBase` through `IDataPortal<T>.FetchAsync()`. We thought the binding of the list would cause the blocking, but we verified it with the following snippet in the `BusinessObjectViewModel<T>`

```
async Task Foo(Func<Task<T>> factory)
{
    var model = await factory();
    Model = model;
}
```

that the blocking happens before `Model` is set. While going through the csla source code, we found that there were no `.ConfigureAwait(false)` applied in the DataPortal access async methods. Is there a reason for this?

**Comment by rockfordlhotka** [comment]
iirc, the use of `ConfigureAwait(false)` on the client side, _especially_ for WinForms/WPF, is problematic because it pushes a lot of work onto the average UI developer. The reason being, that results may (and often do) come back on different threads and/or outside the synchronization context of the UI - forcing you to used UI framework appropriate techniques to shift the result back onto the UI thread.

Basically, again iirc, it prevents the use of data binding in most scenarios.

**Reply by ossendorf-at-hoelscher** [reply]
Doesn't that only apply to BO's which do not get over the wire?
The snippet from above should work as expected because the await there does capture the UI context. But the DataPortals do not have to capture the context _unless_ the used instance is modified. In that case it could be problematic 🤔

**Reply by rockfordlhotka** [reply]
Regardless, the way `await` works is that the compiler creates a state machine, so in this case

```c#
    var model = await factory();
    Model = model;
```

the first line runs and then the method exits, allowing the calling thread (presumably the UI thread) to go on and do whatever, so the UI doesn't lock.

Once the awaited line completes, then the second line executes to set the model - which does happen on the UI thread because the synchronization context was captured.

So yes - there would be a gap in time between the first and second lines running - but in that gap in time, the UI shouldn't be locked because the `await` state machine allowed the method to complete immediately after starting line 1.

**Reply by ossendorf-at-hoelscher** [reply]
But the context flows down the await chain until it hits (ultimately) the http call. But the await there _also_ captures the UI context which means the whole deserialization stuff happens on the UI again instead of a background thread. That causes the UI to lock at that point already instead when returning from the `factory()` call.
Because the context is captured for each `await` in the chain and not at the most top one.

**Reply by rockfordlhotka** [reply]
> Doesn't that only apply to BO's which do not get over the wire?

It should apply in all cases. The client-side data portal _tries_ to make it transparent as to whether you are using a local or remote data portal for your logical server.

My recollection is that there are only a couple scenarios in WinForms (maybe WPF?) where there are some limitations on whether the logical server-side code can or can't use async methods. It has been years. I'm pretty sure the limitations are documented in the _Using CSLA 4_ ebooks.

**Reply by rockfordlhotka** [reply]
> But the context flows down the await chain until it hits (ultimately) the http call. But the await there _also_ captures the UI context which means the whole deserialization stuff happens on the UI again instead of a background thread. That causes the UI to lock at that point already instead when returning from the `factory()` call. Because the context is captured for each `await` in the chain and not at the most top one.

Ahh, good point.

I'm not sure there's a generic way to resolve that - a way that works across all client-side UI frameworks (console, winforms, wpf, aspnetcore, blazor (server/wasm)).

I'm also not sure it is a "CSLA problem", because this would happen with any code that calls an http service unless the UI developer intentionally created a background thread and then shifted the result back onto the UI thread.

That said - if there _was_ a way to generically and reliably solve this across all UI frameworks, that'd be pretty wonderful.

**Reply by ossendorf-at-hoelscher** [reply]
Yeah. The problematic ones are the "in memory" clients. There it _could_ be a problem. But as long as there is no real "await" it should be completed synchronously.
In this case CSLA is abstracting away the http call so I can't say: Yeah I know I don't need the context here and use `ConfigureAwait(false)`.

Maybe a new setting to configure the client side dataportals `ConfigureAwait` parameter through the portal could solve that? 🤔

**Reply by rockfordlhotka** [reply]
We also can't assume http. The data portal can use any network transport, and there are channels for grpc and rabbitmq as well.

**Reply by StefanOssendorf** [reply]
It doesn't matter what kind of transport when it goes over the wire. In that regard the handling is luckily the same.
